### PR TITLE
UserStory #17196: Use Augeas Binaries To get, set and execute commands.

### PR DESCRIPTION
**Pull request by ikramBej**
https://issues.rudder.io/issues/17196

### CODE REVIEW

File: tree/30_generic_methods/augeas_commands.cf [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+#
+# @name Augeas Commands
+# @description Use augeas binaries to execute augtool command(s) directly on the agent. 
+# The generic method will return the command(s) execution output otherwise if the augeas isn't installed on the agent, it will produces an error.
+#
+# @parameter variable_prefix The prefix of the variable name
+# @parameter variable_name   The variable to define, the full name will be variable_prefix.variable_name
+# @parameter commands        The augeas command(s) 
+# @parameter noautoload      Activate the '--no autoload' option if you don't want augeas to charge all the files/lens, it's 'false' by default.
+# @parameter_constraint noautoload "allow_empty_string" : true
```
Comment (amousset) [inline tree/30_generic_methods/augeas_commands.cf:27]: I'd prefer to name the parameter "autoload" with a default to `true`. It is easier to handle "positive" boolean variables name (i.e. `notSomething = true` is more confusing than `Something = false`).

File: tests/acceptance/30_generic_methods/augeas_set.cf [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+#######################################################
+
+bundle agent check
+{
+  classes:
+    
+    "ok_1" expression => "augeas_set_var1_kept.!augeas_set_var1_repaired.!augeas_set_var1_error";
+    "ok_2" expression => "augeas_set_var2_kept.!augeas_set_var2_repaired.!augeas_set_var2_error";
+    "ok_var1" expression => isvariable("prefix.var1");
+    "ok_var2" expression => isvariable("prefix.var2");
```
[the remark this answers is not in the document]
Comment (ikramBej) [reply tests/acceptance/30_generic_methods/augeas_set.cf:51]: I thought as 'set' doesn't return a value, is it necessary to do so? means what we're gonna check exactly?

[the remark this answers is not in the document]
Comment (amousset) [reply tests/acceptance/30_generic_methods/augeas_set.cf:51]: You're right, and the method should not actually have prefix and variable_name parameters in this case.

File: tree/30_generic_methods/augeas_commands.cf (hunk 1 above)
Comment (ikramBej) [reply tree/30_generic_methods/augeas_commands.cf:27]: okay changing this

File: tests/acceptance/30_generic_methods/augeas_set.cf (hunk 2 above)
[the remark this answers is not in the document]
Comment (ikramBej) [reply tests/acceptance/30_generic_methods/augeas_set.cf:51]: okay fixing it

[the remark this answers is not in the document]
Comment (ikramBej) [reply tests/acceptance/30_generic_methods/augeas_set.cf:51]: i'm working on it but i wanna figure out if there is a way we can check if 'execresult' return true or false?

[the remark this answers is not in the document]
Comment (amousset) [reply tests/acceptance/30_generic_methods/augeas_set.cf:51]: You can use `returnszero` instead, which wil define a class if the commands returns zero

[the remark this answers is not in the document]
Comment (ikramBej) [reply tests/acceptance/30_generic_methods/augeas_set.cf]: I have made few changes then made a PR, the test actually works but there just still a little issue with the 'variable_defined', it's always true..

[the remark this answers is not in the document]
Comment (ikramBej) [reply tests/acceptance/30_generic_methods/augeas_set.cf]: as 'returnszero' always return true

**Comment by amousset** [comment]
The tests that modify global values on the system should go into the "unsafe" directory, so in this case the "set" test that modifies global /etc/hosts file.

**Comment by amousset** [comment]
The get test is missing a failure case

File: tree/30_generic_methods/variable_string_from_augeas.cf [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#####################################################################################
+#
+# @name Variable string from Augeas 
+# @description Use Augeas binary to call Augtool command and get a node's label value.
+# @documentation There is two ways you can use this method, either you simply provide a path to the node label as a parameter or you can specify a file associated with a lens then the regular path. 
+# The generic method will get a node's label vaule otherwise if the augeas isn't installed on the agent, it will produces an error.
+#
+# #### Two uses cases examples:
```
Comment (amousset) [inline tree/30_generic_methods/variable_string_from_augeas.cf:24]: Could you add an explanation about the difference of behavior when providing file and lens or not?

Comment (ikramBej) [reply tree/30_generic_methods/variable_string_from_augeas.cf:24]: sure!

File: tree/30_generic_methods/file_augeas_commands.cf [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+# @class_parameter variable_name
+
+
+bundle agent file_augeas_commands(variable_prefix, variable_name, commands, autoload)
+{
+ vars:       
+     "old_class_prefix" string => canonify("file_augeas_commands_${variable_name}");
+     "promisers" slist => { @{this.callers_promisers}, cf_null }, policy => "ifdefined";
+     "class_prefix" string => canonify(join("_", "promisers"));
+     "args" slist => {"${variable_prefix}", "${variable_name}", "${commands}", "${noautoload}"};
```
Comment (amousset) [inline tree/30_generic_methods/file_augeas_commands.cf:39]: noautoload -> autoload

**Pull request merged**